Add support for sqlalchemy any operator to query arrays

bodik opened the pull request:
Add support for sqlalchemy any operator to query arrays. "any" filter models having value in attribute of array type, "not_all" filter models not having value in attribute of array. Adds postgresql specific tests/models for querying arrays.

bodik committed:
Add support for filtering on arrays with postgresql sqlalchemy ANY operator.

* "any"/"not_any" operators allow to filter query by having/not having value in attribute of type array
* corresponding tests and test models

bodik commented:
rebased to master

bodik commented:
hi, any chance to have this reviewed or even merged ?

juliotrigo commented on test/interface/test_filters.py:
```diff
@@ -69,6 +74,15 @@ def multiple_quxs_inserted(session):
+def multiple_corges_inserted(session, is_postgresql):
+    if is_postgresql:
+        corge_1 = Corge(id=1, name='name_1', tags=['foo'])
+        corge_2 = Corge(id=2, name='name_2', tags=['bar', 'baz'])
```
It could be useful to add a couple of rows like the following:

```python
corge_3 = Corge(id=3, name='name_3', tags=['bar', 'foo', 'baz'])
corge_4 = Corge(id=4, name='name_4', tags=[])
```

This way we cover all the cases: we have a `corge` where all tags are `foo`, one where some tags are `foo`, one where none of the tags are `foo` and one with no tag values.

By doing that, it'll be reflected in the tests what happens when only some of the values are `foo` and how we expect `any` and `not_all` to behave.

bodik replied:
fixed with b0e1b4f47c2e746a92229b5e034a98230a4a14e4

juliotrigo commented on test/interface/test_filters.py:
```diff
@@ -1151,3 +1165,36 @@ def test_complex_using_tuples(self, session):
+class TestApplyArrayFilters:
+
+    @pytest.mark.usefixtures('multiple_corges_inserted')
+    def test_value_in_array(self, session, is_postgresql):
```
We could rename this test to something like `test_any_value_in_array` to reflect the operator's action.

bodik replied:
fixed with b0e1b4f

juliotrigo commented on test/interface/test_filters.py:
```diff
@@ -1151,3 +1165,36 @@ def test_complex_using_tuples(self, session):
+        assert result[0].tags == ['foo']
+
+    @pytest.mark.usefixtures('multiple_corges_inserted')
+    def test_value_not_in_array(self, session, is_postgresql):
```
We could rename this to something like `test_not_all_values_in_array` to reflect the operator's action.

bodik replied:
fixed with b0e1b4f

juliotrigo commented on README.rst:
```diff
@@ -303,6 +303,14 @@ This is the list of operators that can be used:
+^^^^^^^^^^^^^
+
+PostgreSQL specific operators can be used to filter queries based on presence
+(``any``) or absence (``not_all``) value in model attribute of ARRAY type.
```
Does `not_all` describe the absence of a value or that not all the values are equal to the one we're filtering by?

I have suggested the addition of another test (in a different comment) to check for this case.

bodik replied:
enhanced testcases from 0318f4d revealed different behavior between sqla1.0 and sqla1.1+, more specifically a behavior of `~` / `not_` operator. The implementation was completely changed to support all sqla APIs correctly.

juliotrigo commented on README.rst:
```diff
@@ -303,6 +303,14 @@ This is the list of operators that can be used:
+any / not_all
+^^^^^^^^^^^^^
+
+PostgreSQL specific operators can be used to filter queries based on presence
```
We could rephrase some parts of this description:
* "based on the presence"
* "of a value in a model attribute of type ``ARRAY``."

bodik replied:
Implementation has been changed and documentation rephrased.

juliotrigo reviewed (commented):
Nice addition @bodik !

I have left a few comments.

bodik commented:
based on the comments so far I have:
* refactored tests naming as suggested
* enhanced tests cases and in turn fixed/re-implemented not_all to not_any operator as the original implementation did not work correctly under sqla1.0 for new tests
* rephrased the documentation
* squashed and rebased to master

juliotrigo commented on README.rst:
```diff
@@ -305,6 +305,14 @@ This is the list of operators that can be used:
+any / not_any
+^^^^^^^^^^^^^
+
+PostgreSQL specific operators allows to filter queries on ARRAY type columns.
```
We could rephrase this description like this, to also highlight ``ARRAY`` as code:

```
PostgreSQL specific operators allow to filter queries on columns of type ``ARRAY``.
Use ``any`` to filter if a value is present in an array and ``not_any`` if it's not.
```

bodik replied:
done

juliotrigo reviewed (commented):
Thanks for the changes!

It's good that we have identified the issues and that tests cover all the different cases now.

One final suggestion about the documentation.

bodik commented:
not sure why travis build is not linked but docs were updated

juliotrigo reviewed (approved):
LGTM 👍

Not sure why Travis is not building. There was another PR merged to `master` an hour ago and maybe that's related. Please try to update your branch with `master` and push the changes to this branch.

juliotrigo reviewed (approved):
LGTM 👍
